A d.ts file for magic leap api exposures

Palmer-JC opened the pull request:
- needed for typescript references / IDE intellisense
- From the page https://github.com/webmixedreality/exokit/blob/master/docs/magicleap.md
- A few questions shown as comments
- There are 2 MLEyeTracker ( one commented out right now )
- Javadoc style comments improve the IDE intellisense, but was shooting for accuracy first.

Also, not sure this is in a place where it will be found.  It is metadata, but for the user, not the building of exokit.

modulesio commented on metadata/exokitMagicLeap.d.ts:
```diff
@@ -0,0 +1,123 @@
+    normalArray : Float32Array;
+    indexBuffer : WebGLBuffer;
+    indexArray : Uint16Array;
+    count : number;  // docs had as 'Number'.  Is this an object or a literal?
```
It should be a primitive `number`. In general we use primitives when possible.

modulesio commented on metadata/exokitMagicLeap.d.ts:
```diff
@@ -0,0 +1,123 @@
+
+declare class MLHandUpdate {
+    hand : whichHand;
+    pointer : MLTransform; // what does the '?' mean?
```
`?` generally means optional in the doc.

modulesio commented on metadata/exokitMagicLeap.d.ts:
```diff
@@ -0,0 +1,123 @@
+/*
+ * This is compented out until the duplicate class name can be resolved.
+
+declare class MLEyeTracker {
```
The existing one with `fixation` should be correct.

modulesio commented:
Maybe this interface file (with comments) could replace the `.md`?

Palmer-JC commented:
Added the comments from the doc.  Kind of by 'definition' this obsoletes the .md file.

Something does look odd about eye now that I removed the second definition.  Shouldn't `MLEyeTracker.eyes` be an array to handle both eyes?

modulesio commented on metadata/exokitMagicLeap.d.ts:
```diff
@@ -1,49 +1,170 @@
+     * The individual eye locations and statuses. Note that this does not include the eye fixation (cursor);
+     * that is contained in fixation.
+     */
     eyes : MLEye;
```
Yeah, this should be `MLEye[2]`

Palmer-JC commented:
Ok, eyes is now an array.  In process of actually making sure it works in an IDE.  Using Netbeans with a Typescript add-in, I am not getting those entry points quite working yet.  I used interfaces so `browser` could be extended, but I do not think `browser` is in the main d.ts file to be extended yet.

The other stuff seems coming thru fine.  Paragraphs do not really in Netbeans, but not sure if this is universal.  No real harm in as programs which generate documentation from files like this may well use this.

Current screen shot.  Stay tuned:
![intellisense](https://user-images.githubusercontent.com/7916131/54615050-ce832700-4a33-11e9-9672-06637bb42d54.jpg)

Palmer-JC commented:
I just made `browser` a module & ` magicLeap` a class within & it works.

Am just going down every object & the intellisense working good.  But starting randomly re-assigning things which probably do nothing.  There is a `readonly` attribute which can be added for code enforcement.  I set MLMeshUpdate.count to `readonly` & transpiler errors.

![readonly](https://user-images.githubusercontent.com/7916131/54626980-b36ee200-4a48-11e9-875f-a0aff2cf9221.jpg)

Do you wish to indicate which should have readonly, or just edit yourself after I have done my last PR?

Palmer-JC commented:
Also there is `ReadonlyArray`, e.g. `wrist : ReadonlyArray<Float32Array>;`

Some of this might seem obvious, but does produce an error during writing when encountered.  Much better than at debug.

Palmer-JC commented:
Probably pretty important to let one know what the position / rotations represent.  y-up or z-up?  Left or right handed?

modulesio commented:
It's right-handed, y-up, z-toward.

Palmer-JC commented:
Think it is ready.  Also tested with Microsoft Visual Code (different from Microsoft Visual Studio 201x).  This is a multi-platform editor used by many.  Did as expected.  Normally d.ts files are generated, so hand built ones need more testing.

![using MSC](https://user-images.githubusercontent.com/7916131/54696842-f097aa80-4b02-11e9-8f76-878c7897c828.JPG)

For future, this is the typescript used to exercise:
```typescript
export class MagicLeap {
    public static isOnMagicLeap() : boolean {
        return browser.magicLeap !== null;
    }

    public test() : void {
        let mesher : MLMesher = browser.magicLeap.RequestMeshing();
        mesher.onmesh = (arg: MLMeshUpdate[]) => { arg[0].count; };
        mesher.destroy();

        let meshUpdate : MLMeshUpdate;
        meshUpdate.id.replace('w', 'W');
        meshUpdate.type = 'unchanged';
        meshUpdate.positionBuffer.is32Bits;
        meshUpdate.positionArray.buffer.slice(3);
        meshUpdate.normalBuffer.capacity + 4;
        meshUpdate.normalArray.BYTES_PER_ELEMENT;
        meshUpdate.indexBuffer.references;
        meshUpdate.indexArray[3] - 6.3;
        meshUpdate.count > 5;

        let eyeTracker: MLEyeTracker = browser.magicLeap.RequestEyeTracking();
        eyeTracker.fixation.position.byteOffset;
        eyeTracker.eyes[0].blink;
        eyeTracker.destroy();

        let planeTracker: MLPlaneTracker = browser.magicLeap.RequestPlaneTracking();
        planeTracker.onplane = (arg: MLPlaneUpdate[]) => { arg[0].id.substring(4); };
        planeTracker.destroy();

        let planeUpdate : MLPlaneUpdate;
        planeUpdate.id.substr(6);
        planeUpdate.position[2].toFixed(3);
        planeUpdate.rotation.subarray(1);
        planeUpdate.size.set([2,3], 2);

        let handTracker : MLHandTracker = browser.magicLeap.RequestHandTracking();  
        handTracker.onhands = (arg : MLHandUpdate[]) => { arg[0].pointer[0] + 3; };
        handTracker.ongesture = (arg: MLGestureUpdate) => { arg.hand.charAt(2); };  
        handTracker.destroy();     

        let handUpdate : MLHandUpdate;
        handUpdate.hand === 'left';
        handUpdate.pointer.position;
        handUpdate.grip.rotation[3];
        handUpdate.rotation[0];
        handUpdate.wrist.length;
        handUpdate.wrist[2].buffer;
        handUpdate.wrist[2][1] + 34.2;
        handUpdate.fingers.length;
        handUpdate.fingers[0].length;
        handUpdate.fingers[0][0].length;
        handUpdate.fingers[0][0][0].buffer;
        handUpdate.fingers[0][0][0][1] - 9;
        handUpdate.gesture === 'pinch';

        let gestureUpdate : MLGestureUpdate;
        gestureUpdate.hand === 'left';
        gestureUpdate.gesture === 'pinch';
        gestureUpdate.position[2] * -8.6;
        gestureUpdate.rotation.byteOffset;

        let transform : MLTransform;
        transform.position[0] * transform.rotation[3];

        let eye : MLEye;
        eye.blink ? eye.position[0] : eye.rotation[2];

        let image : HTMLImageElement;
        let imageTracker: MLImageTracker = browser.magicLeap.RequestImageTracking(image, 4);
        imageTracker.ontrack = (arg? : MLTrackingUpdate) => { if (arg) arg.position.buffer.byteLength; };

        let trackingUpdate : MLTrackingUpdate;
        (trackingUpdate.position[1] / trackingUpdate.rotation[2]) + trackingUpdate.size;
    }
}
```
Enjoy!

modulesio commented:
Awesome, thanks! This should be a lot better than the existing `.md`.
